### ISSUE TITLE
Suggestion: OpenFaas with preload function

### ISSUE THREAD

**Issue by teo-chenglim** [openfaas/faas#1017]
Suggest the watchdog shall able to pre-load some code and execute the remaining code at run-time. It is a bit like the python generator, yield.

Let's say a function total running time is 30 seconds, pre-load take 20 seconds then it goes to run-time code 10 seconds. Other than time saving, it also help because we fetch the weight from file (disk IO) and this can be ease if you have this design.

Another reason why you should have this is because Data Science/Machine Learning/AI are all base on this model. We load pre-trained weight (the 20 seconds) and do prediction on run-time (10 seconds)

Please consider this suggestion. Thank you.

**Comment by ivanayov** [comment]
Derek add label: proposal

**Comment by ivanayov** [comment]
Hi @teo-chenglim ,

Thank you for your proposal.
My concern about it is that this means to keep a state for a function.

Can you share an example with preload that keeps the function stateless, if you have such in mind?
What I imagine is that the watchdog should run the preload once and then keep the state for every future invocation.

Ivana

**Comment by alexellis** [comment]
Hi @teo-chenglim the best way to do this would be to use one of the of-watchdog templates. You can find these by typing in `faas-cli template store list -v`.

You may want to try `python3-flask` for instance which allows for pre-loading of a model into memory.

Does this help? 

Alex

**Comment by alexellis** [comment]
I'll ping @dkozlov who is experienced with this method too and has made use of GPUs with OpenFaaS for AI/ML.

I'll close the support issue now, but please keep commenting.

Alex

**Comment by teo-chenglim** [comment]
Actually second try it works.

```bash
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ curl -s https://faas.amaris.ai/function/heavy1 -d "faas"
awesome
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ cat function/faasinit.py 
faas = "awesome\r\n"
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ cat template/python3-flask
python3-flask/       python3-flask-armhf/ 
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ cat template/python3-flask
python3-flask/       python3-flask-armhf/ 
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ cat template/python3-flask/index.py 
# Copyright (c) Alex Ellis 2017. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.

from flask import Flask, request
from function import handler
#from gevent.wsgi import WSGIServer
from gevent.pywsgi import WSGIServer

try:
  from function.faasinit import * # This is the heavy loading scripts faasinit.py
except ImportError:
  pass

app = Flask(__name__)

@app.before_request
def fix_transfer_encoding():
    """
    Sets the "wsgi.input_terminated" environment flag, thus enabling
    Werkzeug to pass chunked requests as streams.  The gunicorn server
    should set this, but it's not yet been implemented.
    """

    transfer_encoding = request.headers.get("Transfer-Encoding", None)
    if transfer_encoding == u"chunked":
        request.environ["wsgi.input_terminated"] = True

@app.route("/", defaults={"path": ""}, methods=["POST", "GET"])
@app.route("/<path:path>", methods=["POST", "GET"])
def main_route(path):
    #ret = handler.handle(request.get_data())
    #ret = handler.handle(faas)
    ret = handler.handle(request.get_data(), faas)
    return ret

if __name__ == '__main__':
    #app.run(host='0.0.0.0', port=5000, debug=False)

    http_server = WSGIServer(('', 5000), app)
    http_server.serve_forever()
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ cat function/handler.py 
def handle(req, faas):
    """handle a request to the function
    Args:
        req (str): request body
    """

    return faas
chenglim@chenglim-GL503VM:/amaris/faas/heavy1$ 

```

**Comment by alexellis** [comment]
You can get this template via `faas template store` command.

If you'd like to customise the base that can be done too in a forked Git repo. You then do `faas template pull https://`

As for the additional arguments, that would be nice, but we have users already so we would need to use a new name for it. 

Alex

**Comment by teo-chenglim** [comment]
Hi @alexellis I got the proof of concept working with Keras resnet50  (https://keras.io/applications/).
It is not feasible to pass param using "**kwargs" as initial thought. So any added param still need to touch 3 files now, namely "index.py", "faasinit.py" and "handler.py". But it is so many moving parts.

So instead of creating a new template, i would write an tutorial how i do it when i got more time.

I believe with my example it is yet another good show case of openfaas which is much more extensible than many other serverless frameworks in term of flexibility, customisable and easy to use. 

```bash
$ curl -s https://endpoint/function/flask-preload -d "faas"
[["n02504458", "African_elephant", 0.5097971558570862], ["n01871265", "tusker", 0.45040571689605713], ["n02504013", "Indian_elephant", 0.03866980969905853]]
```

**Comment by alexellis** [comment]
@teo-chenglim do you think you could write this up on the OpenFaaS blog or perhaps on your own?

Alex

**Comment by loretoparisi** [comment]
@alexellis this is a very interesting thread. Has been addressed in some way? In the past we had several concerns on model preloading in python AI apis...it turned out that using `tornado` (that has async io) it worked ok in production environments (while in my opinion `flask` cannot be recommended for production systems...).
Thank you!
